sound/soc/sof/intel/hda-stream.c

```diff
@@ -223,14 +224,14 @@ int hda_dsp_stream_put(struct snd_sof_dev *sdev, int direction, int stream_tag)
 		if (s->direction == direction && s->stream_tag == stream_tag) {
 			s->opened = false;
 			found = true;
-		} else if (s->direction == SNDRV_PCM_STREAM_CAPTURE) {
+		} else if (s->direction == SNDRV_PCM_STREAM_CAPTURE && !d0i3_compatible) {
```

Comment:
I think we need to store information on whether stream is d0i3_compatible to "struct sof_intel_hda_stream" in stream_get(), and then in stream_put, use this information.

Now above you are going through all remaining open streams, but you don't know whether other streams are d0i3_compatible or not, so this won't quite work. For s->direction the logic works as we have stored the direction information for all stream. I think we need to do same for d0i3_compatible.

Reply:
nice catch @kv2019i . fixed now.